### PR TITLE
Implemented linear gradient fill in Android

### PR DESCRIPTION
**Pull request by tepamid**
Partial implementation of "TODO(6352048): Support gradients etc.": linear gradient fill implemented.

We implemented a custom Color Picker based on ReactART.

screen shot 2016-12-19 at 12 27 09 pm

### CODE REVIEW

**Comment by hramos** [comment]
Can you provide a bit more context? As your original comment at the top will be included if this PR is merged, I'd like to have a more detailed explanation of what this PR does.

File: ReactAndroid/src/main/java/com/facebook/react/views/art/ARTShapeShadowNode.java [hunk 1]
```diff
@@ -182,6 +184,34 @@ protected boolean setupFillPaint(Paint paint, float opacity) {
...
+              int g = (int) (255 * mFillColor[5 + 4*i + 1]);
+              int b = (int) (255 * mFillColor[5 + 4*i + 2]);
+              int a = (int) (255 * mFillColor[5 + 4*i + 3]);
+              colors[i] = (a << 24) | (r << 16) | (g << 8) | (b << 0);
+            }
+          }
+          paint.setShader(new LinearGradient(x0, y0, x1, y1, colors, positions, Shader.TileMode.CLAMP));
+          break;
         default:
           // TODO(6352048): Support gradients etc.
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/views/art/ARTShapeShadowNode.java]: Should this TODO be removed?

File: ReactAndroid/src/main/java/com/facebook/react/views/art/ARTShapeShadowNode.java [hunk 2]
```diff
@@ -182,6 +184,34 @@ protected boolean setupFillPaint(Paint paint, float opacity) {
               (int) (mFillColor[2] * 255),
               (int) (mFillColor[3] * 255));
           break;
+        case 1: // linear gradient
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/views/art/ARTShapeShadowNode.java]: Why is `1` a linear gradient? Can you use some constant instead or add a comment with link to docs explaining why `1` should be a linear gradient? This is very non-obvious.

File: ReactAndroid/src/main/java/com/facebook/react/views/art/ARTShapeShadowNode.java [hunk 3]
```diff
@@ -182,6 +184,34 @@ protected boolean setupFillPaint(Paint paint, float opacity) {
...
+          float y0 = mFillColor[2];
+          float x1 = mFillColor[3];
+          float y1 = mFillColor[4];
+          int stops = (mFillColor.length - 5) / 5;
+          int[] colors = null;
+          float[] positions = null;
+          if (stops > 0) {
+            colors = new int[stops];
+            positions = new float[stops];
+            for(int i=0; i<stops; i++) {
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/views/art/ARTShapeShadowNode.java]: nit: Space after `for`

File: ReactAndroid/src/main/java/com/facebook/react/views/art/ARTShapeShadowNode.java [hunk 4]
```diff
@@ -182,6 +184,34 @@ protected boolean setupFillPaint(Paint paint, float opacity) {
...
+          if (mFillColor.length < 5) {
+            FLog.w(ReactConstants.TAG, 
+              "[ARTShapeShadowNode setupFillPaint] expects 5 elements, received " 
+              + mFillColor.length);
+            return false;
+          }
+          float x0 = mFillColor[1];
+          float y0 = mFillColor[2];
+          float x1 = mFillColor[3];
+          float y1 = mFillColor[4];
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/views/art/ARTShapeShadowNode.java]: What are these 4 variables? Add a comment explaining what they represent. Are they the coordinates of the start and end of the gradient? If yes, why not name them `gradientStartX` etc.?

**Comment by isilher** [comment]
@tepamid are you planning to work on this? And if not, do you mind if I pick it up?

**Comment by tepamid** [comment]
Fixed code-style issues and added necessary comments. @mkonicek @kmagiera please review.

File: ReactAndroid/src/main/java/com/facebook/react/views/art/ARTShapeShadowNode.java [hunk 5]
```diff
@@ -175,15 +185,54 @@ protected boolean setupFillPaint(Paint paint, float opacity) {
...
+          if (mFillColor.length < 5) {
+            FLog.w(ReactConstants.TAG, 
+              "[ARTShapeShadowNode setupFillPaint] expects 5 elements, received " 
+              + mFillColor.length);
+            return false;
+          }
+          float gradientStartX = mFillColor[1];
+          float gradientStartY = mFillColor[2];
+          float gradientEndX = mFillColor[3];
+          float gradientEndY = mFillColor[4];
```
Comment (isilher) [inline ReactAndroid/src/main/java/com/facebook/react/views/art/ARTShapeShadowNode.java]: Each of these coordinates needs to be multiplied by `mScale` to support high density screens:

```
render() {
    const linearGradient1 = new LinearGradient(['red', 'green', 'red'], 0, 0, 100, 0)
    const linearGradient2 = new LinearGradient(['red', 'green', 'red'], 0, 0, 0, 200)
    const linearGradient3 = new LinearGradient(['red', 'green', 'red'], 0, 0, 300, 300)

    return (
      <View>
        <Surface width={100} height={100}>
          <Shape x={0} y={0} d="M10 10 H 90 V 90 H 10 L 10 10" fill={linearGradient1} />
        </Surface>
        <Surface width={200} height={200}>
          <Shape x={0} y={0} d="M10 10 H 190 V 190 H 10 L 10 10" fill={linearGradient2} />
        </Surface>
                <Surface width={300} height={300}>
          <Shape x={0} y={0} d="M10 10 H 290 V 290 H 10 L 10 10" fill={linearGradient3} />
        </Surface>
      </View>
    )
  }
```

**Without mScale, comparing MotoG3 to iOS simulator**
img_2843

**With mScale, comparing motoG3 to iOS simulator**
img_4248 2

**Comment by tepamid** [comment]
@kmagiera ok, I'll work on all suggestions during the upcoming weekend.

**Comment by tepamid** [comment]
I made the requested changes and verified them as part of my application.

**Comment by isilher** [comment]
I verified your changes in my application, all works well.

**Comment by kmagiera** [comment]
@facebook-github-bot shipit

cc: @ericvicenti

**Comment by kmagiera** [comment]
@facebook-github-bot shipit

**Pull request closed without merging**